tests_and_analysis/test/euphonic_test/test_spectrum2d.py

```diff
@@ -431,3 +432,60 @@ def test_get_bin_centres_with_invalid_data_shape_raises_value_error(
             spec2d.get_bin_centres()
 
 
+class TestKinematicAngles:
+
+    @pytest.mark.parametrize(
+        'angle_range, expected',
+        [((0, np.pi / 2), (1, 0)),
+         ((np.pi / 2, 0), (1, 0)),
+         ((0, 1.4 * np.pi), (1, -1)),
+         ((-2.25 * np.pi, -2.5 * np.pi), (np.sqrt(2) / 2, 0))
+         ])
+    def test_cos_range(self, angle_range, expected):
+        from euphonic.spectra import _get_cos_range
+        cos_limits = _get_cos_range(angle_range)
+        assert cos_limits == pytest.approx(expected)
+
+
+class TestKinematicConstraints:
+
+    @pytest.mark.parametrize(
+        'kwargs, spectrum2d_file, constrained_file',
+        [({'e_i': 300 * ureg('1/cm')},
+          'NaCl_band_yaml_dos_map.json',
+          'NaCl_constrained_ei_300_cm.json'),
+         ({'e_i': 100 * ureg('1/cm'), 'angle_range': (20, 275)},
+          'NaCl_band_yaml_dos_map.json',
+          'NaCl_constrained_ei_100_cm_angle_20_275.json'),
+         ({'e_i': 30 * ureg('meV'), 'angle_range': (-20, 30)},
+          'NaCl_band_yaml_dos_map.json',
+          'NaCl_constrained_ei_30_meV_angle_-20_30.json'),
+         ({'e_f': 32 * ureg('1/cm'), 'angle_range': (45., 135.)},
+          'NaCl_band_yaml_dos_map.json',
+          'NaCl_constrained_ef_32_cm_angle_45_135.json'),
+         ({'e_f': 32 * ureg('1/cm'), 'angle_range': (-135., -45.)},
+          'NaCl_band_yaml_dos_map.json',
+          'NaCl_constrained_ef_32_cm_angle_45_135.json')
+         ])
+    def test_kinematic_constraints(self, kwargs,
+                                   spectrum2d_file, constrained_file):
+        spec2d = get_spectrum2d(spectrum2d_file)
+        ref_spec2d = get_spectrum2d(constrained_file)
+
+        constrained_spec2d = apply_kinematic_constraints(spec2d, **kwargs)
+        check_spectrum2d(constrained_spec2d, ref_spec2d, equal_nan=True)
+
+    @pytest.mark.parametrize('json_file, kwargs, expected',
+                             [('NaCl_band_yaml_dos_map.json',
+                               {'e_i': 10 * ureg('1/cm'),
+                                'e_f': 20 * ureg('1/cm')},
+                               ValueError),
+                              ('NaCl_band_yaml_dos_map.json', {}, ValueError),
+                              ('example_spectrum2d.json',
+                               {'e_i': 20 * ureg('1/cm')},
+                               ValueError)])
```

Comment:
This tests the case where y_data is invalid ('dimensionless') but not x_data. We could do with an example where x_data is in invalid units too, but there don't seem to be any in the current test data. Maybe change the x units of 'example_xbin_edges_spectrum2d.json' to something silly? I don't think that would have an adverse affect on any other tests.